parsers/html.c

```diff
@@ -8,35 +8,405 @@
 };
 
-/*
-*   FUNCTION DEFINITIONS
-*/
+typedef enum {
+	KEYWORD_h1,
+	KEYWORD_h2,
+	KEYWORD_h3,
+	KEYWORD_a,
+	KEYWORD_script,
+	KEYWORD_style,
+	KEYWORD_name
+} keywordId;
+
+static const keywordTable HtmlKeywordTable[] = {
+	{"h1", KEYWORD_h1},
+	{"h2", KEYWORD_h2},
+	{"h3", KEYWORD_h3},
+	{"a", KEYWORD_a},
+	{"script", KEYWORD_script},
+	{"style", KEYWORD_style},
+	{"name", KEYWORD_name},
+};
+
+typedef enum {
+	TOKEN_EOF,
+	TOKEN_NAME,			/* tag and attribute names */
+	TOKEN_STRING,		/* single- or double-quoted attribute value */
+	TOKEN_TEXT,
+	TOKEN_TAG_START,	/* <  */
+	TOKEN_TAG_START2,	/* </ */
+	TOKEN_TAG_END,		/* >  */
+	TOKEN_TAG_END2,		/* /> */
+	TOKEN_EQUAL,
+	TOKEN_COMMENT,
+	TOKEN_OTHER
+} tokenType;
+
+typedef struct {
+	tokenType type;
+	vString *string;
+} tokenInfo;
+
+
+static int Lang_html;
+
+
+static void readTag (tokenInfo *token, vString *text);
+
+
+static void readTokenText (tokenInfo *const token, boolean collectText)
+{
+	int c;
+	int lastC = 'X';  /* whatever non-space character */
+
+	vStringClear (token->string);
+
+getNextChar:
+
+	c = getcFromInputFile ();
+
+	switch (c)
+	{
+		case EOF:
+			token->type = TOKEN_EOF;
+			break;
+
+		case '<':
+		{
+			ungetcToInputFile (c);
+			token->type = TOKEN_TEXT;
+			break;
+		}
+
+		default:
+			if (collectText)
+			{
+				if (isspace (c))
+					c = ' ';
+				if (c != ' ' || (c == ' ' && lastC != ' '))
+				{
+					if (collectText)
+						vStringPut (token->string, c);
+					lastC = c;
+				}
+			}
+
+			goto getNextChar;
+	}
+}
+
+static void readToken (tokenInfo *const token, boolean skipComments)
+{
+	int c;
+
+	vStringClear (token->string);
+
+getNextChar:
+
+	c = getcFromInputFile ();
+	while (isspace (c))
+		c = getcFromInputFile ();
+
+	switch (c)
+	{
+		case EOF:
+			token->type = TOKEN_EOF;
+			break;
+
+		case '<':
+		{
+			int d = getcFromInputFile ();
+
+			if (d == '!')
+			{
+				d = getcFromInputFile ();
+				if (d == '-')
+				{
+					d = getcFromInputFile ();
+					if (d == '-')
+					{
+						int e = ' ';
+						int f = ' ';
+						do
+						{
+							d = e;
+							e = f;
+							f = getcFromInputFile ();
+						}
+						while (f != EOF && ! (d == '-' && e == '-' && f == '>'));
+
+						if (skipComments)
+							goto getNextChar;
+						else
+						{
+							token->type = TOKEN_COMMENT;
+							break;
+						}
+					}
+				}
+				ungetcToInputFile (d);
+				token->type = TOKEN_OTHER;
+			}
+			else if (d == '?')
+				token->type = TOKEN_OTHER;
+			else if (d == '/')
+				token->type = TOKEN_TAG_START2;
+			else
+			{
+				ungetcToInputFile (d);
+				token->type = TOKEN_TAG_START;
+			}
+			break;
+		}
+		case '/':
+		{
+			int d = getcFromInputFile ();
+			if (d == '>')
+				token->type = TOKEN_TAG_END2;
+			else
+			{
+				ungetcToInputFile (d);
+				token->type = TOKEN_OTHER;
+			}
+			break;
+		}
+		case '>':
+		{
+			token->type = TOKEN_TAG_END;
+			break;
+		}
+
+		case '=':
+			token->type = TOKEN_EQUAL;
+			break;
+
```

Comment:
For the case `'>'`, you put `{` and `}` around the statements after the case label.
However, the case `'='`, you don't.  Could you do the same (putting or not putting)?